docs/zkapps/tutorials/08-custom-tokens.mdx

```diff
@@ -0,0 +1,143 @@
+---
+title: "Tutorial 8: Custom Tokens"
+hide_title: true
+sidebar_label: "Tutorial 8: Custom Tokens"
+---
+
+:::info
+
+Please note that zkApp programmability is not yet available on Mina Mainnet, but
+zkApps can now be deployed to Berkeley Testnet.
+
+:::
+
+# Tutorial 8: Custom Tokens
+
+## Overview
+
+In this tutorial, we will discuss how to create custom tokens.
+
+Mina comes with native support for custom tokens. Each account on Mina can also have tokens associated with it.
+
+To create a new token, one creates a smart contract, which becomes the manager for the token, and uses that contract to set the rules around how the token can be mint, burned, and sent.
+
+The manager account may also set a token symbol for its token.
+
+In this tutorial, we will review smart contract code that creates and manages new tokens. You can find a full example, including these smart contracts and code to launch and interact with them, [here](https://github.com/es92/zkApp-examples/tree/main/08-custom-tokens/src).
+
+You can also find a more extensive example, including all the ways to interact with token smart contracts, [here](https://github.com/o1-labs/snarkyjs/blob/main/src/lib/token.test.ts).
+
+## Basic Token Example
+
+To create a token manager smart contract, we will create a normal smart contract, whose methods call special functions that manipulate tokens.
+
+You can find a full copy of this file [here](https://github.com/es92/zkApp-examples/blob/main/08-custom-tokens/src/BasicTokenContract.ts).
+
+First, we will bring in imports and setup the structure for our smart contract:
+
+```ts
+  1 import {
+  2   Field,
+  3   SmartContract,
+  4   state,
+  5   State,
+  6   method,
+  7   DeployArgs,
+  8   Permissions,
+  9   UInt64,
+ 10   PublicKey,
+ 11   Signature,
+ 12 } from 'snarkyjs';
+ 13
+ 14 const tokenSymbol = 'MYTKN';
+ 15
+ 16 export class BasicTokenContract extends SmartContract {
+ 17   @state(UInt64) totalAmountInCirculation = State<UInt64>();
+ 18
+ 19   deploy(args: DeployArgs) {
+ 20     super.deploy(args);
+ 21
+ 22     this.setPermissions({
+ 23       ...Permissions.default(),
+ 24       editState: Permissions.proofOrSignature(),
+ 25       setTokenSymbol: Permissions.proofOrSignature(),
+ 26     send: Permissions.proofOrSignature(),
+ 27     receive: Permissions.proofOrSignature(),
+ 28     });
+ 29   }
+ 30 }
+```
+
+We include a single state variable, totalAmountInCirculation,  which we will use to track how many tokens exist.
+
+Next we’ll add an init method, and a method that mints tokens:
+
+```ts
+ 31   @method init() {
+ 32     this.tokenSymbol.set(tokenSymbol);
+ 33     this.totalAmountInCirculation.set(UInt64.zero);
+ 34   }
+ 35
+ 36   @method mint(
+ 37     receiverAddress: PublicKey,
+ 38     amount: UInt64,
+ 39     adminSignature: Signature
+ 40   ) {
+ 41     let totalAmountInCirculation = this.totalAmountInCirculation.get();
+ 42 this.totalAmountInCirculation.assertEquals(totalAmountInCirculation);
```

Comment:
For `assertEquals` and other assert methods, users can now pass in a string to be logged if the constraint fails, which acts similarly to `require` in Ethereum. This would be a great place to demonstrate it's usage by logging something like `totalAmountInCirculation does not match` or something like this. Could you fill in these assert methods with helpful error messages?

Reply:
I buy that in general, but do we want to do that on checks like this, that are just checking the call to `.get()`? Maybe Tutorial 2 would be a good place for it, I think there's an assertEqual there we could say something about how the secret that was passed wasn't correct that could be useful?